File: app/src/main/java/org/evcs/android/util/ViewUtils.java
package org.evcs.android.util;

import android.app.Activity;
import android.content.Context;
import android.os.Parcel;
import android.os.Parcelable;
import androidx.annotation.NonNull;

import android.text.Spannable;
import android.text.SpannableString;
import android.text.TextPaint;
import android.text.style.URLSpan;
import android.text.style.UnderlineSpan;
import android.view.View;
import android.widget.TextView;

import java.util.concurrent.atomic.AtomicInteger;

/**
 * Utilities for Android views
 */
public class ViewUtils {

    private ViewUtils() {}

    /**
     * Determines and sets a view visibility depending on a flag.
     *
     * @param view Target {@link View}.
     * @param show If true the visibility will be {@link View#VISIBLE}, else {@link View#GONE}.
     */
    public static void setViewVisibility(@NonNull View view, boolean show) {
        view.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    /**
     * Returns the width of the screen in pixels
     * @param context the context of the view
     */
    public static float getWindowWidth(Context context) {
        return ((Activity) context).getWindowManager().getDefaultDisplay().getWidth();
    }

    /**
     * Removes the underlines for a {@link TextView} with auto generated URL links for email or phones.
     *
     * @param textView TextView to remove underlines from
     */
    public static void stripUnderlines(TextView textView) {
        Spannable s = new SpannableString(textView.getText());
        URLSpan[] spans = s.getSpans(0, s.length(), URLSpan.class);
        for (URLSpan span: spans) {
            int start = s.getSpanStart(span);
            int end = s.getSpanEnd(span);
            s.removeSpan(span);
            span = new URLSpanNoUnderline(span.getURL());
            s.setSpan(span, start, end, 0);
        }
        textView.setText(s);
    }

    public static void addUnderlines(TextView textView) {
        SpannableString content = new SpannableString(textView.getText());
        content.setSpan(new UnderlineSpan(), 0, content.length(), 0);
        textView.setText(content);
    }

    /**
     * Span to remove underline
     */
    private static class URLSpanNoUnderline extends URLSpan {

        private static final Parcelable.Creator<URLSpanNoUnderline> CREATOR =
            new Creator<URLSpanNoUnderline>() {
            @Override
            public URLSpanNoUnderline createFromParcel(Parcel source) {
                return new URLSpanNoUnderline(source.readString());
            }

            @Override
            public URLSpanNoUnderline[] newArray(int size) {
                return new URLSpanNoUnderline[size];
            }
        };

        public URLSpanNoUnderline(String url) {
            super(url);
        }

        @Override
        public void updateDrawState(@NonNull TextPaint ds) {
            super.updateDrawState(ds);
            ds.setUnderlineText(false);
        }
    }

    public static void showIfTheresScrolling(@NonNull final View shadow, @NonNull final View scrollView) {
        scrollView.addOnLayoutChangeListener(new View.OnLayoutChangeListener() {
            @Override
            public void onLayoutChange(View view, int i, int i1, int i2, int i3, int i4, int i5, int i6, int i7) {
                boolean canScroll = scrollView.canScrollVertically(1) || scrollView.canScrollVertically(-1);
                shadow.setVisibility(canScroll ? View.VISIBLE : View.GONE);
            }
        });
    }
}

File: app/src/main/java/org/evcs/android/util/ViewUtils.kt
package org.evcs.android.util

import android.app.Activity
import android.content.Context
import android.widget.TextView
import android.text.Spannable
import android.text.SpannableString
import android.text.style.URLSpan
import android.text.style.UnderlineSpan
import android.text.TextPaint
import android.view.View
import android.view.ViewGroup
import androidx.core.view.ViewCompat
import androidx.core.view.isVisible

/**
 * Utilities for Android views
 */
object ViewUtils {
    /**
     * Determines and sets a view visibility depending on a flag.
     *
     * @param view Target [View].
     * @param show If true the visibility will be [View.VISIBLE], else [View.GONE].
     */
    fun View.setVisibility(show: Boolean) {
        this.visibility = if (show) View.VISIBLE else View.GONE
    }

    fun View.setParentVisibility(show: Boolean) {
        (this.parent as View).isVisible = show
    }

    fun TextView.showOrHide(text: String?) {
        isVisible = text != null
        setText(text)
    }

    /**
     * Returns the width of the screen in pixels
     * @param context the context of the view
     */
    fun getWindowWidth(context: Context): Float {
        return (context as Activity).windowManager.defaultDisplay.width.toFloat()
    }

    fun View.setMargins(left: Int, top: Int, right: Int, bottom: Int) {
        val lp = ViewGroup.MarginLayoutParams(layoutParams ?: ViewGroup.LayoutParams(-2, -2))
        lp.setMargins(left, top, right, bottom)
        layoutParams = lp
    }

    fun Context.getStatusBarHeight(): Int {
        val resId =
                resources.getIdentifier("status_bar_height", "dimen", "android")
        return if (resId > 0) resources.getDimensionPixelSize(resId) else 0
    }

    /**
     * Workaround to keep the adjust resize behaviour without ruining the toolbar with fitsSystemWindows
     */
    fun setAdjustResize(top: View) {
        ViewCompat.setOnApplyWindowInsetsListener(top) { v, insets ->
            v.setPadding(0, 0, 0, insets.systemWindowInsetBottom)
            insets.consumeSystemWindowInsets()
        }
    }

    /**
     * Removes the underlines for a [TextView] with auto generated URL links for email or phones.
     *
     * @param textView TextView to remove underlines from
     */
    fun stripUnderlines(textView: TextView) {
        val s: Spannable = SpannableString(textView.text)
        val spans = s.getSpans(0, s.length, URLSpan::class.java)
        for (span in spans) {
            val start = s.getSpanStart(span)
            val end = s.getSpanEnd(span)
            s.removeSpan(span)
            val span2 = URLSpanNoUnderline(span.url)
            s.setSpan(span2, start, end, 0)
        }
        textView.text = s
    }

    fun addUnderlines(textView: TextView) {
        val content = SpannableString(textView.text)
        content.setSpan(UnderlineSpan(), 0, content.length, 0)
        textView.text = content
    }

    fun showIfTheresScrolling(shadow: View, scrollView: View) {
        scrollView.addOnLayoutChangeListener { _, _, _, _, _, _, _, _, _ ->
            val canScroll = scrollView.canScrollVertically(1) || scrollView.canScrollVertically(-1)
            shadow.visibility = if (canScroll) View.VISIBLE else View.GONE
        }
    }

    /**
     * Span to remove underline
     */
    class URLSpanNoUnderline(url: String?) : URLSpan(url) {
        override fun updateDrawState(ds: TextPaint) {
            super.updateDrawState(ds)
            ds.isUnderlineText = false
        }

        /*object {
            private val CREATOR: Creator<URLSpanNoUnderline> =
                object : Creator<URLSpanNoUnderline> {
                    override fun createFromParcel(source: Parcel): URLSpanNoUnderline {
                        return URLSpanNoUnderline(source.readString())
                    }

                    override fun newArray(size: Int): Array<URLSpanNoUnderline?> {
                        return arrayOfNulls(size)
                    }
                }
        }*/
    }
}